product/src/main/java/com/yas/product/service/ProductService.java

```diff
@@ -184,15 +186,34 @@ private void validateIfProductWithSkuOrGtinOrSlugExist(String slug,
                                                            String gtin,
                                                            String sku) {
         if (isProductWithSlugAvailable(slug)) {
-            throw new DuplicatedException(Constants.ErrorCode.SLUG_ALREADY_EXISTED, slug);
+            throw new DuplicatedException(Constants.ErrorCode.SLUG_ALREADY_EXISTED_OR_DUPLICATED, slug);
         }
 
         if (isProductWithGtinAvailable(gtin)) {
-            throw new DuplicatedException(Constants.ErrorCode.GTIN_ALREADY_EXISTED, gtin);
+            throw new DuplicatedException(Constants.ErrorCode.GTIN_ALREADY_EXISTED_OR_DUPLICATED, gtin);
         }
 
         if (isProductWithSkuAvailable(sku)) {
-            throw new DuplicatedException(Constants.ErrorCode.SKU_ALREADY_EXISTED, sku);
+            throw new DuplicatedException(Constants.ErrorCode.SKU_ALREADY_EXISTED_OR_DUPLICATED, sku);
+        }
+    }
+
+    private void validateIfProductSkuOrGtinOrSlugDuplicated(List<ProductVariationPostVm> variations) {
+        Set<String> seenSlugs = new HashSet<>();
+        Set<String> seenSkus = new HashSet<>();
+        Set<String> seenGtins = new HashSet<>();
+        for (ProductVariationPostVm variation : variations) {
+            if (!seenSlugs.add(variation.slug())) {
+                throw new DuplicatedException(Constants.ErrorCode.SLUG_ALREADY_EXISTED_OR_DUPLICATED, variation.slug());
+            }
+
+            if (!seenGtins.add(variation.gtin())) {
+                throw new DuplicatedException(Constants.ErrorCode.GTIN_ALREADY_EXISTED_OR_DUPLICATED, variation.gtin());
+            }
+
+            if (!seenSkus.add(variation.sku())) {
+                throw new DuplicatedException(Constants.ErrorCode.SKU_ALREADY_EXISTED_OR_DUPLICATED, variation.sku());
+            }
```

Comment:
use `Map` instead of `Set` to optimize code

Reply:
Hi @minhtranq-nashtechglobal

Thank you for your suggestion. I just wanted to clarify something: if we were to use Map instead of Set, we would still need to initialize 3 separate maps to check for duplicates (one for each property). Is that correct?

I’m considering the trade-offs and would appreciate your thoughts on this.

Thanks again for your input.

Reply:
yes, correct

Reply:
I think I’ll stick with Set for this implementation. Thank you for taking the time to review this issue!